rasa/core/lock.py

```diff
@@ -0,0 +1,158 @@
+import json
+import logging
+import time
+from collections import deque
+from typing import Text, Optional, Union, Deque, Dict, Any
+
+logger = logging.getLogger(__name__)
+
+
+class Ticket:
+    def __init__(self, number: int, expires: float):
+        self.number = number
+        self.expires = expires
+
+    def has_expired(self) -> bool:
+        return time.time() > self.expires
```

Comment:
what happens in the case where someone has a ticket, the action is running very long and the ticket expires in the meantime?

Reply:
nothing, once the lock has been acquired it's locked. `has_expired()` is checked only _before_ acquisition

Reply:
ok but in that case what happens if an instance gets restarted while it is holding a lock? your message sounds like the lock would never be available anymore.